### PR TITLE
core(emulation): set client-hints metadata when spoofing the UA

### PR DESCRIPTION
**Pull request by paulirish**
We've discovered that some webservers misbehave when they get a mobile useragent but empty clienthints headers.

And since a while ago... when we set [`Emulation.setUserAgentOverride`](https://chromedevtools.github.io/devtools-protocol/tot/Emulation/#method-setUserAgentOverride) .. it appropriately sets the useragent string, but since no `userAgentMetadata` was provided, all those values are undefined. As client hints are being adopted, this leads to weird and brokenish scenarios. (depending on how defensive folks are configuring their ua handling on the webserver side).

We'll want to make sure clienthint headers are appropriately populated.

I went for making a change that minimizes impact on people who configure lighthouse with custom UA. I think we can revisit exactly how to expose those fields. I'd go for something less verbose than pptr's choice of [replicating](https://pptr.dev/#?product=Puppeteer&version=v11.0.0&show=api-pagesetuseragentuseragent-useragentmetadata) the exact same payload.

-------

* handy test page: https://browserleaks.com/client-hints
* devtools's constants: https://github.com/ChromeDevTools/devtools-frontend/blob/main/front_end/panels/network/NetworkConfigView.ts#L369
* if you're wondering [why 'Not a Brand' is in your UA-CH](https://wicg.github.io/ua-client-hints/#grease).

### CODE REVIEW

File: lighthouse-core/lib/emulation.js [hunk 1]
```diff
@@ -16,15 +16,56 @@ const NO_CPU_THROTTLE_METRICS = {
...
+ * @param {LH.Config.Settings['formFactor']} formFactor
+ * @return {LH.Crdp.Emulation.SetUserAgentOverrideRequest['userAgentMetadata']}
+ */
+function parseUseragentIntoMetadata(userAgent, formFactor) {
+  const match = userAgent.match(/Chrome\/([\d.]+)/); // eg 'Chrome/(71.0.3577.0)'
+  const fullVersion = (match && match[1]) || '99.0.1234.0';
+  const version = fullVersion.split('.', 1)[0];
+  const brands = [
+    {brand: 'Chromium', version},
+    {brand: 'Google Chrome', version},
```
Comment (connorjclark) [inline lighthouse-core/lib/emulation.js:30]: wdyt of adding a entry for LH version here?

Comment (connorjclark) [reply lighthouse-core/lib/emulation.js:30]: and if we do that, we should randomly sort the array, to support the GREASE stuffs (or will chrome do that for us in the request headers?)

File: lighthouse-core/lib/emulation.js [hunk 2]
```diff
@@ -16,15 +16,56 @@ const NO_CPU_THROTTLE_METRICS = {
...
 async function emulate(session, settings) {
   if (settings.emulatedUserAgent !== false) {
+    const userAgent = /** @type {string} */ (settings.emulatedUserAgent);
     await session.sendCommand('Network.setUserAgentOverride', {
-      userAgent: /** @type {string} */ (settings.emulatedUserAgent),
+      userAgent,
+      userAgentMetadata: parseUseragentIntoMetadata(userAgent, settings.formFactor),
     });
   }
   // See devtools-entry for one usecase for disabling screenEmulation
```
Comment (connorjclark) [inline lighthouse-core/lib/emulation.js:71]: needs tests, esp. for the case where the UA is user-provided and doesn't even have a chrome version in it

Comment (connorjclark) [inline lighthouse-core/lib/emulation.js:71]: needs tests, esp. for the case where the UA is user-provided and doesn't even have a chrome version in it

File: lighthouse-core/lib/emulation.js [hunk 3]
```diff
@@ -16,15 +16,56 @@ const NO_CPU_THROTTLE_METRICS = {
...
 
+/**
+ * @param {string} userAgent
+ * @param {LH.Config.Settings['formFactor']} formFactor
+ * @return {LH.Crdp.Emulation.SetUserAgentOverrideRequest['userAgentMetadata']}
+ */
+function parseUseragentIntoMetadata(userAgent, formFactor) {
+  const match = userAgent.match(/Chrome\/([\d.]+)/); // eg 'Chrome/(71.0.3577.0)'
+  const fullVersion = (match && match[1]) || '99.0.1234.0';
+  const version = fullVersion.split('.', 1)[0];
```
Comment (connorjclark) [inline lighthouse-core/lib/emulation.js:27]:
```suggestion
  const [version] = fullVersion.split('.', 1);
```

😎

File: lighthouse-core/lib/emulation.js (hunk 1 above)
Comment (paulirish) [reply lighthouse-core/lib/emulation.js:32]: > wdyt of adding a entry for LH version here?

done

> and if we do that, we should randomly sort the array, to support the GREASE stuffs (or will chrome do that for us in the request headers?)

seems like fun.. and its jsut adding a `.sort(_ => Math.random() - 0.5)`  

... but we add some non-determinimism without much gain so i think we should put it off. leave the greasing to the big boiz.

**Review by connorjclark — APPROVED** [review]

**Pull request merged**